Closed connection detection when in socket_loop

yrashk opened the pull request:
We needed this so that when we're in keepalive mode we can instantaneously detect that the client is gone

yrashk committed:
socket_loop should be able to detect if the remote party is gone (tcp_closed) and quit immediately, because is no sense in trying to send anything to the closed socket.

yrashk committed:
Added a handler for ssl_closed in addition to tcp_closed

yrashk committed:
Improve matching for tcp_closed/ssl_closed in socket_loop

ostinelli commented:
afa e0d531b is concerned.

socket_loop is not receiving TCP data, it is normally receiving data only from the handler Loop function, spawned for every client request. plus, receiving TCP data in the socket_loop will result in calling the _Else case, and not treating it appropriately

am i missing something?

yrashk commented:
It has worked for us so far, but it would probably not work if the remote side will keep sending something. Can we possibly find a way to hack misultin to be able to handle both situations?

ostinelli commented:
sure thing. i believe this should be addressed more generally, on some lists there have been requests to get the CLOSED event even on the http handler loop.

let's think about it, since right now this proposal will result in dropping some incoming requests from a client sending multiple requests at the same time, and this cannot be accepted.

yrashk commented:
We're still trying to attack this problem at different angles. Did you have a good reason for that _Else clause in `socket_loop/3` in the first place as it seems to protect from "fools" — from data sent into it by some code that was evil enough to get the request process pid out — which should normally be only accessed by misultin_req functions only?

The reason why I am asking is that if we get rid of _Else clause, actual received tcp data won't hurt as it will be filtered out until we're back in `request/2` (in keep_alive case). And even if we would have that kind of evil code sending unsupported messages, it will eventually hit _Other clause in `request/2`. The behaviour will be more or less of the same kind, as while in `socket_loop/3`, it won't make `socket_loop/3` do anything (which is what happens on _Else, it just ignores and flushes that message) but will be flushed in `request/2`, albeit somewhat differently — by printing a different message and actually closing the socket. But shouldn't evil code sending something totally unexpected be penalized for its behaviour anyway? You can't make a fully fool-proof code anyway.

yrashk committed:
Dropped _Else clause in socket_loop as it prevents things from failing if some evil code tries to feed socket_loop with unsupported messages. This way tcp messages other than tcp_closed and ssl_closed are saved until request/2 is called (if it is a keep_alive connection) so if client decides to reuse a connection (which is possible in HTTP/1.1) it won't have data eaten by _Else clause.

Instead of being too forgiving, its connection will be rather forcefully dropped (if it was a keep_alive connection) next time request/2 is called, which is a right way to penalize an unruly code.

yrashk commented:
I've added a commit that follows up on the logic I suggested above.

ostinelli commented:
this still doesn't solve the error. you are calling setops here:
https://github.com/yrashk/misultin/commit/e0d531bb0187e02a7ab8c0eec08666fd9cd0efc6#L0R298

this leads to two problems.
1. except for the tcp_closed event, you are not treating tcp data. so any incoming tcp data which does not match this event will cause a no matching clause error here:
   https://github.com/yrashk/misultin/blob/master/src/misultin_http.erl#L357
2. where you to reintroduce the _Else clause that you took away, the tcp data would be lost in there, and then you'll get to https://github.com/ostinelli/misultin/blob/master/src/misultin_http.erl#L76
   where setops is called again, and the tcp data has definitely been lost into the _Else statement

thus i cannot accept this pull request right now.

what i'm still missing is why you need to detect the "client is gone". are you using comet? otherwise, the handle_loop will simply send data into a non-existing socket and then exit.

yrashk commented:
It will not cause a no matching clause error, it will make Erlang filter message out until we're out of that receive so it will be received by consequently called request/2 in case of keepalive, which is what we want; and no data will be lost

http://ndpar.blogspot.com/2010/11/erlang-explained-selective-receive.html :

If none of the messages in the mailbox matches, then the process is suspended and will be rescheduled for execution the next time a new message is put in the mailbox. Note that when a new message arrives, the messages in the save queue are not rematched; only the new message is matched.

I need to detect "client is gone" situation for socket.io-erlang (kind of like comet)

ostinelli commented:
indeed correct for my point 1. point 2 still remains: tcp data will be lost and trigger this warning:
https://github.com/ostinelli/misultin/blob/master/src/misultin_http.erl#L96

i think that for comet applications there might be a need to more seriously refactor this whole http_handle attitude in this particular case. the arch of misultin is oriented at treating single requests one at a time, except for the websockets part of course - but that's a different story.

for instance: if you do have a client flooding with http requests [malconfigured? dos attempt?], this will lead to receiving a portion of an http request that will be lost, causing concatenated errors.

one more [important] thing: tcp_closed being catched by socket_loop does _not_ mean that the _separate_ process handle_loop is aware of this happening. this _will_ create handle_loop to stay up as a zombie process [memleak]: so you need a way to pass this information to handle_loop too. that's why i said there might be a need for a deeper refactoring.

however, your point is well taken and is interesting. i've had similar requests on other boards.

i'm also trying to figure out typical user cases where this could be problematic.

what do you think.

yrashk commented:
Potential data loss can be easily avoided by setting http packet parser option in that setopts in socket_loop, no? (Good catch, by the way!). I will update my patch shortly.

yrashk committed:
Avoid missing http data passed through the socket_loop/3

yrashk commented:
Updated patch to avoid losing data

yrashk commented:
I am not sure I fully agree on the zombie point. Wouldn't you get this situation in any keepalive when something goes wrong? I might be missing something, tired.

ostinelli commented:
hu? don't see why this should change anything, except when receiving POST data, the socket already is in {packet, http} mode.

yrashk commented:
I am not sure why do you think the data will be lost. With selective receive any tcp data that is not tcp_closed or ssl_closed all those messages will make it into request/2 receive loop, no?

ostinelli commented:
yes, and therefore end up here:
https://github.com/ostinelli/misultin/blob/master/src/misultin_http.erl#L96

which will, by the way, close the connection:
https://github.com/ostinelli/misultin/blob/master/src/misultin_http.erl#L99

yrashk commented:
but since we'll be receiving http packets, wouldn't one of this match instead:

```
    {SocketMode, Sock, {http_request, Method, Path, Version}} ->
    {SocketMode, Sock, {http_error, "\r\n"}} ->
    {SocketMode, Sock, {http_error, "\n"}} ->
    {http, Sock, {http_error, _Other}}  ->
```

?

(Disclaimer: I know next to nothing about developing web servers)

yrashk commented:
I still don't quite undertand how passing tcp data through selective receive further down to the next round of request/2 is going to hurt exactly? What exactly am I missing?

ostinelli commented:
added.
